File: examples/buttons/server/rails/test/integration/checkout_flow_test.rb
# frozen_string_literal: true

require "test_helper"

# The whole hop-by-hop path over real HTTP, with a stubbed NWC wallet:
#
#   POST /shop/orders          mint ONE order row, priced from shop_products
#   POST /openreceive/checkouts  -> authorize -> amount_for -> an invoice
#   POST /openreceive/payments/check -> settlement -> on_paid -> claim_paid!
#   GET  /shop/orders/:id      re-read the row; downloads have unlocked
#   GET  .../downloads/:sku    the artwork
#
# The first test in this file is the gate the whole plan rests on: whether
# `config.authorize` can read a SIGNED COOKIE off the bare ActionDispatch::Request
# the hook is handed.
class CheckoutFlowTest < ActionDispatch::IntegrationTest
  include ButtonShopTestSetup

  JSON_HEADERS = { "CONTENT_TYPE" => "application/json" }.freeze

  setup do
    @wallet = FakeWallet.new
    stub_openreceive!(@wallet)
    seed_catalog!
    ensure_pack_assets!
  end

  test "authorize reads the signed cookie off the bare request" do
    order = create_order!(sku: "safety-orange", quantity: 1)

    # The browser that placed the order may check out.
    post "/openreceive/checkouts",
         params: JSON.generate(reference: order.fetch("reference")), headers: JSON_HEADERS
    assert_response :created
  end

  test "another browser cannot check out somebody else's order" do
    order = create_order!(sku: "safety-orange", quantity: 1)
    reference = order.fetch("reference")

    # A different browser: no cookie at all.
    reset!
    post "/openreceive/checkouts",
         params: JSON.generate(reference: reference), headers: JSON_HEADERS
    assert_response :forbidden

    # And an unsigned cookie carrying the real owner's uuid is still nobody.
    owner_id = ShopOrder.find(reference).shop_user_id
    cookies[ShopIdentity::COOKIE] = owner_id
    post "/openreceive/checkouts",
         params: JSON.generate(reference: reference), headers: JSON_HEADERS
    assert_response :forbidden

    assert_equal 0, OpenReceivePayment.count
  end

  test "checkout mints an invoice, settles, and unlocks the download" do
    order = create_order!(sku: "safety-orange", quantity: 2)
    reference = order.fetch("reference")

    checkout = create_checkout!(reference)
    payment_hash = checkout.fetch("payment_hash")
    assert_match(/\A[0-9a-f]{64}\z/, payment_hash)
    assert checkout.fetch("bolt11")
    # 2 × $1.00 at the static $50,000/BTC test price is exactly 4000 sats.
    assert_equal 4_000_000, checkout.fetch("amount_msats")

    assert_equal "pending", check_payment!(reference, payment_hash).fetch("status")

    # Literal paths: after a request to the mounted engine the integration
    # session would prefix named routes with the engine's /openreceive mount.
    get "/shop/orders/#{reference}/downloads/safety-orange"
    assert_response :forbidden

    @wallet.settle!(payment_hash, at: Time.now.to_i)
    # The durable openreceive_meta gate collapses rapid polls to one wallet scan
    # per 2s interval; the payer's next poll after the interval wins it.
    travel 3.seconds
    assert_equal "settled", check_payment!(reference, payment_hash).fetch("status")

    stored = ShopOrder.find(reference)
    assert_equal ShopOrder::PAID, stored.state
    assert_equal payment_hash, stored.payment_hash
    assert stored.paid_at.present?

    get "/shop/orders/#{reference}"
    assert_response :success
    assert_equal "/shop/orders/#{reference}/downloads/safety-orange",
                 JSON.parse(response.body).fetch("items").sole.fetch("download_path")

    get "/shop/orders/#{reference}/downloads/safety-orange"
    assert_response :success
    assert_equal "image/webp", response.media_type
  end

  test "the download survives the browser being closed and reopened" do
    order = create_order!(sku: "signal-red", quantity: 1)
    reference = order.fetch("reference")
    cookie = cookies[ShopIdentity::COOKIE]

    payment_hash = create_checkout!(reference).fetch("payment_hash")
    @wallet.settle!(payment_hash, at: Time.now.to_i)
    travel 3.seconds
    check_payment!(reference, payment_hash)

    # A new session is a new browser process. The signed cookie is what it
    # brings back, and it is the whole of what makes the order still theirs.
    reset!
    cookies[ShopIdentity::COOKIE] = cookie

    get "/shop/orders/#{reference}/downloads/signal-red"
    assert_response :success
    assert_equal 1, ShopUser.count
  end

  test "reconciliation settles a pending attempt without browser polling" do
    reference = create_order!(sku: "midnight-navy", quantity: 1).fetch("reference")
    payment_hash = create_checkout!(reference).fetch("payment_hash")
    @wallet.settle!(payment_hash, at: Time.now.to_i)

    assert_equal 1, OpenReceive.reconcile!.length

    assert_equal ShopOrder::PAID, ShopOrder.find(reference).state
    assert_equal "settled", OpenReceivePayment.find_by(payment_hash: payment_hash).status
  end

  test "a settled order appears in the public feed, to a visitor who never bought anything" do
    reference = create_order!(sku: "classic-black", quantity: 1).fetch("reference")
    buyer_ref = ShopOrder.find(reference).shop_user.public_ref
    payment_hash = create_checkout!(reference).fetch("payment_hash")
    @wallet.settle!(payment_hash, at: Time.now.to_i)
    assert_equal 1, OpenReceive.reconcile!.length

    reset!
    get "/shop/recent_orders"

    assert_response :success
    row = JSON.parse(response.body).fetch("orders").sole
    assert_equal buyer_ref, row.fetch("buyer")
    assert_equal 500, row.fetch("total_cents")
    assert_equal "Classic Black", row.fetch("items").sole.fetch("name")
    refute_includes response.body, reference
  end

  test "rate limiting stamps the client IP on committed attempts" do
    reference = create_order!(sku: "plain-white", quantity: 1).fetch("reference")
    create_checkout!(reference)

    assert_equal ["127.0.0.1"], OpenReceivePayment.pluck(:client_ip).uniq
  end

  test "the SPA shell is not cached so webpack rebuilds reach the browser" do
    get "/"

    assert_response :success
    assert_includes response.headers["Cache-Control"], "no-store"
    assert_match %r{/packs-test/js/buttons-[0-9a-f]+\.js}, response.body
  end

  private

  def create_order!(sku:, quantity:)
    post "/shop/orders", params: { items: [{ sku: sku, quantity: quantity }] }, as: :json
    assert_response :created
    JSON.parse(response.body)
  end

  def create_checkout!(reference)
    post "/openreceive/checkouts",
         params: JSON.generate(reference: reference), headers: JSON_HEADERS
    assert_response :created
    JSON.parse(response.body).fetch("checkout")
  end

  def check_payment!(reference, payment_hash)
    post "/openreceive/payments/check",
         params: JSON.generate(reference: reference, payment_hash: payment_hash),
         headers: JSON_HEADERS
    assert_response :success
    JSON.parse(response.body)
  end
end

File: examples/buttons/server/laravel/tests/Feature/CheckoutFlowTest.php
<?php

declare(strict_types=1);

namespace Tests\Feature;

use App\Models\ShopOrder;
use App\Support\CatalogSeed;
use App\Support\Visitor;
use Illuminate\Testing\TestResponse;
use Tests\TestCase;

/**
 * The whole bridge, end to end, with the wallet stubbed: bootstrap → order →
 * the mounted engine mints against the fake wallet → the wallet settles →
 * payments/check runs onPaid inside the settlement transaction → the order
 * row is paid and the download unlocks. Another browser sees none of it.
 */
final class CheckoutFlowTest extends TestCase
{
    private string $visitorCookie = '';

    protected function setUp(): void
    {
        parent::setUp();
        CatalogSeed::apply();
        // The bootstrap mints the visitor; its cookie is what every later request carries.
        $bootstrap = $this->getJson('/shop/bootstrap');
        $bootstrap->assertOk();
        $cookie = $bootstrap->getCookie(Visitor::COOKIE);
        self::assertNotNull($cookie);
        $this->visitorCookie = (string) $cookie->getValue();
    }

    /** @param array<string, mixed> $body */
    private function asVisitor(string $method, string $uri, array $body = []): TestResponse
    {
        // getCookie() decrypted the value; the test client encrypts it again on the
        // way out — and only sends cookies on a JSON request with withCredentials().
        return $this->withCredentials()->withCookie(Visitor::COOKIE, $this->visitorCookie)->json($method, $uri, $body);
    }

    public function testBootstrapCarriesTheServerPricedCatalogAndTheMountPrefix(): void
    {
        $response = $this->asVisitor('GET', '/shop/bootstrap');
        $response->assertOk()->assertHeader('Cache-Control', 'no-store, private');
        self::assertSame('/openreceive', $response->json('shop.openreceive_prefix'));
        self::assertCount(6, $response->json('shop.catalog'));
        self::assertSame(['sku' => 'safety-orange', 'name' => 'Safety Orange', 'price_cents' => 100, 'image_url' => '/images/openreceive-safety-orange-button.webp'], $response->json('shop.catalog.0'));
        self::assertMatchesRegularExpression('/^[0-9a-f-]{36}$/', $response->json('shop.visitor.public_ref'));
    }

    public function testACartIsPricedFromTheProductRowsNeverFromTheWire(): void
    {
        $created = $this->asVisitor('POST', '/shop/orders', ['items' => [
            ['sku' => 'safety-orange', 'quantity' => 2, 'price_cents' => 1],
            ['sku' => 'classic-black', 'quantity' => 1],
            ['sku' => 'not-a-product', 'quantity' => 5],
            ['sku' => 'signal-red', 'quantity' => 99],
        ]]);
        $created->assertStatus(201);
        self::assertSame('awaiting_payment', $created->json('state'));
        // 2 × $1.00 + $5.00 + 10 (clamped) × $10.00
        self::assertSame(10_700, $created->json('total_cents'));
        self::assertSame('107.00', $created->json('total_amount'));
        self::assertSame('OpenReceive buttons: Safety Orange ×2, Classic Black, Signal Red ×10', $created->json('description'));
        self::assertNull($created->json('items.0.download_path'));

        $this->asVisitor('POST', '/shop/orders', ['items' => []])->assertStatus(422)->assertJsonPath('error', 'Your cart is empty.');
    }

    public function testTheLightningCheckoutSettlesThroughTheThreeHooks(): void
    {
        $reference = (string) $this->asVisitor('POST', '/shop/orders', ['items' => [['sku' => 'safety-orange', 'quantity' => 1]]])->json('reference');

        // The mounted engine: authorize (this visitor owns the order) → amountFor ($1.00 → 2,000 sats) → mint.
        $checkout = $this->asVisitor('POST', '/openreceive/checkouts', ['reference' => $reference]);
        $checkout->assertStatus(201);
        $hash = (string) $checkout->json('checkout.payment_hash');
        self::assertSame(str_repeat('0', 63).'1', $hash);
        self::assertSame(2_000_000, $checkout->json('checkout.amount_msats'));
        self::assertSame('OpenReceive button: Safety Orange', $checkout->json('description'));
        self::assertStringStartsWith('lnbcopenreceive', (string) $checkout->json('checkout.bolt11'));

        // Still unpaid: no download.
        self::assertSame('awaiting_payment', $this->asVisitor('GET', "/shop/orders/{$reference}")->json('state'));
        $this->asVisitor('GET', "/shop/orders/{$reference}/downloads/safety-orange")->assertStatus(403);

        // The wallet reports the payment; the next status poll settles it and runs onPaid ONCE.
        $this->wallet->settleInvoice(['payment_hash' => $hash]);
        $check = $this->asVisitor('POST', '/openreceive/payments/check', ['reference' => $reference, 'payment_hash' => $hash]);
        $check->assertOk();
        self::assertSame('settled', $check->json('status'));

        $order = $this->asVisitor('GET', "/shop/orders/{$reference}");
        self::assertSame('paid', $order->json('state'));
        self::assertSame("/shop/orders/{$reference}/downloads/safety-orange", $order->json('items.0.download_path'));
        self::assertSame($hash, ShopOrder::query()->findOrFail($reference)->payment_hash);

        $download = $this->asVisitor('GET', "/shop/orders/{$reference}/downloads/safety-orange");
        $download->assertOk();
        self::assertSame('image/webp', $download->headers->get('Content-Type'));

        // A second payment can never re-fulfill: the engine refuses a new checkout under a settled reference.
        $this->asVisitor('POST', '/openreceive/checkouts', ['reference' => $reference])->assertStatus(409);

        // The public feed shows the paid order, attributed to the PUBLIC handle, and never the order id.
        $feed = $this->getJson('/shop/recent_orders');
        $feed->assertOk()->assertHeader('Cache-Control', 'max-age=10, public');
        self::assertSame(1, $feed->json('totals.paid_orders'));
        self::assertSame(1, $feed->json('totals.buttons_sold'));
        self::assertSame('Safety Orange', $feed->json('orders.0.items.0.name'));
        self::assertStringNotContainsString($reference, $feed->getContent());
    }

    public function testAnotherBrowserCannotSeeOrPayForThisOrder(): void
    {
        $reference = (string) $this->asVisitor('POST', '/shop/orders', ['items' => [['sku' => 'safety-orange', 'quantity' => 1]]])->json('reference');

        // No cookie: a NEW visitor, who does not own the order. 404, never 403 — do not confirm the id exists.
        $this->defaultCookies = [];
        $this->withCredentials = false;
        $this->getJson("/shop/orders/{$reference}")->assertStatus(404);
        // And the engine's authorize hook says no before any wallet call.
        $stranger = $this->postJson('/openreceive/checkouts/prepare', ['reference' => $reference]);
        $stranger->assertStatus(403);
        self::assertSame('FORBIDDEN', $stranger->json('code'));
        self::assertSame([], $this->wallet->listInvoices());
    }

    public function testTheArtworkIsServedFromTheOneImagesDirectory(): void
    {
        $this->get('/images/openreceive-safety-orange-button.webp')->assertOk()->assertHeader('Content-Type', 'image/webp');
        $this->get('/images/nope.webp')->assertStatus(404);
    }
}
